### PR TITLE
fix prompt tunning + deepspeed zero3 + checkpoint_saving hang issue

### PR DESCRIPTION
**Pull request by sywangyi**
# What does this PR do?

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by sywangyi** [comment]
Hi @younesbelkada. When I use deepspeed zero3 and prompt tuning to finetune large model. find training hang after saving checkpoint.
 the prompt-tuning has prompt_encoder forward in model save_pretrained. only rank0 call the prompt_encoder forward in deepspeed zero3 is not enough. all the rank should call prompt_encoder forward.

**Comment by ArthurZucker** [comment]
cc @pacman100 for deepspeed!

**Review by pacman100 — COMMENTED** [review]
Thank you @sywangyi for the fixes and explanation. I agree with the issue and the need to call the `save_pretrained` method on all ranks due to the call to `prompt_encoder(prompt_tokens)`. However, the changes made here make it difficult to maintain. We need to look at a better way to resolve this.

**Comment by sywangyi** [comment]
yes, I think should_save should move to _save. instead of controling if _save is called like _save_tpu

**Comment by yao-matrix** [comment]
@pacman100 , @ArthurZucker what's your suggestion to move on for this issue? any comments on Yi's proposal?

**Comment by sywangyi** [comment]
how about this? @pacman100

**Review by muellerzr — COMMENTED** [review]
Catching up to speed, where I'm a bit lost in is why are we not *just* modifying the chunk when doing `elif self.is_deepspeed_enabled`? Then during `_save` we can check for if we are using deepspeed and do the adjusted logic. 

Or: why does this have to touch every single instance of `_save` essentially. This feels like a regression as `self.args.should_save` should be the main caller for everything

**Comment by sywangyi** [comment]
> the prompt-tuning has prompt_encoder forward in model save_pretrained. only rank0 call the prompt_encoder forward in deepspeed zero3 is not enough. all the rank should call prompt_encoder forward.

thanks for the review!  the root cause of this issue is that the prompt-tuning has prompt_encoder forward in model **save_pretrained**. only rank0 call the prompt_encoder forward in deepspeed zero3 is not enough. all the rank should call prompt_encoder forward. or else it will hang...    so **save_pretrained** in **_save** should be called in all the rank. however. self.args.should_save is only true when local_process_index=0. so I move it to **_save**.

**Comment by muellerzr** [comment]
Again though: this is deepspeed specific based on model sharding/splitting, so we should only modify the code for deepspeed specifically, unless there is a reason not to. (Aka, deepspeed and fsdp specific I imagine).

Am I incorrect in this understanding?

**Comment by sywangyi** [comment]
> Again though: this is deepspeed specific based on model sharding/splitting, so we should only modify the code for deepspeed specifically, unless there is a reason not to. (Aka, deepspeed and fsdp specific I imagine).
> 
> Am I incorrect in this understanding?

yes, deepspeed zero3 and FSDP specific

**Comment by muellerzr** [comment]
Great, let's adjust the logic for that case specifically then, and leave the already existing base logic in-tact then

**Comment by sywangyi** [comment]
@muellerzr  I have updated the PR, could you revisit it?

**Review by muellerzr — APPROVED** [review]
Thanks! LG2M, cc @amyeroberts

File: src/transformers/trainer.py [hunk 1]
```diff
@@ -3443,7 +3440,8 @@ def _save_tpu(self, output_dir: Optional[str] = None):
     def _save(self, output_dir: Optional[str] = None, state_dict=None):
         # If we are executing this function, we are the process zero, so we don't check for that.
         output_dir = output_dir if output_dir is not None else self.args.output_dir
-        os.makedirs(output_dir, exist_ok=True)
+        if self.args.should_save:
+            os.makedirs(output_dir, exist_ok=True)
```
Comment (amyeroberts) [inline src/transformers/trainer.py:3444]: Doesn't this mean the comment above is no longer correct?

File: src/transformers/trainer.py [hunk 2]
```diff
@@ -3443,7 +3440,8 @@ def _save_tpu(self, output_dir: Optional[str] = None):
...
         logger.info(f"Saving model checkpoint to {output_dir}")
```
Comment (amyeroberts) [inline src/transformers/trainer.py:3445]: This is now wrong if `should_save` is False

**Review by amyeroberts — COMMENTED** [review]
Thanks for adding this, but having this fix indicates that there's likely something wrong in how we control our saving logic more generally. 

Having to have lots of `should_save` checks within the `_save` method is odd: the `_save` method should only be called if `should_save` is true OR, at worst, the `_save` method should be directly exited at the start of the call, we shouldn't have to have the whole method's logic conditioned on this flag

File: src/transformers/trainer.py (hunk 1 above)
Comment (sywangyi) [reply src/transformers/trainer.py:3444]: yes, if _save is called in each rank. the comment above is no longer correct.

**Comment by sywangyi** [comment]
> Thanks for adding this, but having this fix indicates that there's likely something wrong in how we control our saving logic more generally.
> 
> Having to have lots of `should_save` checks within the `_save` method is odd: the `_save` method should only be called if `should_save` is true OR, at worst, the `_save` method should be directly exited at the start of the call, we shouldn't have to have the whole method's logic conditioned on this flag

the main aim is that the save_pretrained in _save should be called in every rank in FSDP, DEEPSPEED+prompt tuning case.

File: src/transformers/trainer.py (hunk 2 above)
Comment (amyeroberts) [reply src/transformers/trainer.py:3445]: This wasn't resolved

File: src/transformers/trainer.py [hunk 3]
```diff
@@ -3455,7 +3453,10 @@ def _save(self, output_dir: Optional[str] = None, state_dict=None):
 
             if isinstance(self.accelerator.unwrap_model(self.model), supported_classes):
                 self.accelerator.unwrap_model(self.model).save_pretrained(
-                    output_dir, state_dict=state_dict, safe_serialization=self.args.save_safetensors
+                    output_dir,
+                    state_dict=state_dict,
+                    safe_serialization=self.args.save_safetensors,
+                    is_main_process=self.args.should_save,
                 )
             else:
```
Comment (amyeroberts) [inline src/transformers/trainer.py:3461]: As there's no logic in conditional on `should_save` within this `else` branch, won't this lead to more safetensor savings unexpectedly if the model isn't a pretrained model?

**Review by amyeroberts — COMMENTED** [review]
@sywangyi Thanks for the explanation. I understand the intended logic. My previous comment still stands: we shouldn't need to condition so much of the `_save` logic on the `should_save` flag, it's OK to pass in `is_main_process`, but it doesn't really make sense for saving e.g. the tokenizer. Instead, I would look at: 
* How we define the main process - we might need a different flag e.g. `self._is_main_process`
* Update `should_save` to be true for all ranks in FSDP, DEEPSPEED+prompt tuning case

cc @muellerzr Does this seem reasonable?

File: src/transformers/trainer.py (hunk 3 above)
Comment (sywangyi) [reply src/transformers/trainer.py]: good finding, fix it.

**Comment by sywangyi** [comment]
> @sywangyi Thanks for the explanation. I understand the intended logic. My previous comment still stands: we shouldn't need to condition so much of the `_save` logic on the `should_save` flag, it's OK to pass in `is_main_process`, but it doesn't really make sense for saving e.g. the tokenizer. Instead, I would look at:
> 
> * How we define the main process - we might need a different flag e.g. `self._is_main_process`
> * Update `should_save` to be true for all ranks in FSDP, DEEPSPEED+prompt tuning case
> 
> cc @muellerzr Does this seem reasonable?

_save_tpu has the similar logic. use should_save flag internal.

**Comment by sywangyi** [comment]
also. in `save_pretrained` (https://github.com/huggingface/transformers/blob/main/src/transformers/modeling_utils.py#L2503 or https://github.com/huggingface/peft/blob/main/src/peft/peft_model.py#L294) is_main_process is equal to `should_save`,`should_save` is controlled by https://github.com/huggingface/transformers/blob/main/src/transformers/training_args.py#L2322. not equal to `process_index` == 0. If you see `_save_tpu` logic, it's what I would like to imitate. calling logic is https://github.com/huggingface/transformers/blob/main/src/transformers/trainer.py#L3420. 
function logic is like https://github.com/huggingface/transformers/blob/main/src/transformers/trainer.py#L3460. `should_save` is put in the save function internally.

File: src/transformers/trainer.py [hunk 4]
```diff
@@ -3376,20 +3376,17 @@ def save_model(self, output_dir: Optional[str] = None, _internal_call: bool = Fa
                 version.parse(accelerate_version) > version.parse("0.24.1")
             ):
                 state_dict = self.accelerator.get_state_dict(self.model)
-                if self.args.should_save:
-                    self._save(output_dir, state_dict=state_dict)
+                self._save(output_dir, state_dict=state_dict)
```
Comment (muellerzr) [inline src/transformers/trainer.py:3379]: This still doesn't quite get what Amy was after, and I agree it's the solution we want to look at:

> Update should_save to be true for all ranks in FSDP, DEEPSPEED+prompt tuning case

While yes that's not *quite* how it works for TPU, that difference is important IMO and we should do it this way.

Comment (sywangyi) [reply src/transformers/trainer.py:3379]: currently, should_save is a property of args, and controlled by many factors.
see
image

if we update should_save to be true for all ranks. we need other parameters to replace original should_save to control if we needs to save tokenizer and arguments.
https://github.com/huggingface/transformers/blob/main/src/transformers/trainer.py#L3529-L3533

**Comment by yao-matrix** [comment]
@amyeroberts , do you have any further comments on this? Thx.

**Comment by amyeroberts** [comment]
@yao-matrix Beyond [my comment here](https://github.com/huggingface/transformers/pull/29980#pullrequestreview-2161152984) and @muellerzr's [comment here](https://github.com/huggingface/transformers/pull/29980#discussion_r1699045322), I don't have any else to add, but happy to clarify my point in any way or help iterate on a solution

**Comment by yao-matrix** [comment]
> @yao-matrix Beyond [my comment here](https://github.com/huggingface/transformers/pull/29980#pullrequestreview-2161152984) and @muellerzr's [comment here](https://github.com/huggingface/transformers/pull/29980#discussion_r1699045322), I don't have any else to add, but happy to clarify my point in any way or help iterate on a solution

I read through the whole thread again and will co-work w/ @sywangyi to propose a better solution.

**Pull request closed without merging**